src/Syntax_Server/Oni_Syntax_Server.re

```diff
@@ -0,0 +1,126 @@
+/*
+ Syntax client
+ */
+
+open Oni_Syntax;
+
+module Protocol = Oni_Syntax.Protocol;
+module ClientToServer = Protocol.ClientToServer;
+
+type message =
+  | Log(string)
+  | Message(ClientToServer.t)
+  | Exception;
+
+let start = () => {
+  // A list of pending messages for us to handle
+  let messageQueue: ref(list(message)) = ref([]);
+  // Mutex to guard accessing the queue from multiple threads
+  let messageMutex = Mutex.create();
+
+  let queue = msg => {
+    Mutex.lock(messageMutex);
+    messageQueue := [msg, ...messageQueue^];
+    Mutex.unlock(messageMutex);
+  };
+
+  let flush = () => {
+    Mutex.lock(messageMutex);
+    let result = messageQueue^;
+    messageQueue := [];
+    Mutex.unlock(messageMutex);
+    List.rev(result);
+  };
+
+  let isRunning = ref(true);
+
+  let _runThread: Thread.t =
+    Thread.create(
+      () => {
+        let write = (msg: Protocol.ServerToClient.t) => {
+          Marshal.to_channel(Stdlib.stdout, msg, []);
+          Stdlib.flush(Stdlib.stdout);
+        };
+
+        let log = msg => write(Protocol.ServerToClient.Log(msg));
+
+        let state = ref(State.empty);
+        let map = f => state := f(state^);
+
+        let handleProtocol =
+          fun
+          | ClientToServer.Echo(m) =>
+            write(Protocol.ServerToClient.EchoReply(m))
+          | v => log("Unhandled message: " ++ ClientToServer.show(v));
+
+        let handleMessage =
+          fun
+          | Log(msg) => log(msg)
+          | Exception => log("Exception encountered!")
+          | Message(protocol) => handleProtocol(protocol);
+
+        while (true) {
+          log("Syntax Server - tick");
+
+          // Get pending messages and handle them
+          flush() |> List.iter(handleMessage);
+
+          Unix.sleepf(1.0);
+          // Wait for incoming messages
+          // Handle messages
+          // If any work, flush work
+        };
+      },
+      (),
+    );
+
+  let _readThread: Thread.t =
+    Thread.create(
+      () => {
+        while (isRunning^) {
+          queue(Log("Waiting for message from client..."));
+          let msg: Oni_Syntax.Protocol.ClientToServer.t =
+            Marshal.from_channel(Stdlib.stdin);
+          queue(Log("Got message from client!"));
+          switch (msg) {
+          | exception _ => queue(Exception)
+          | msg => queue(Message(msg))
+          };
```

Comment:
This looks odd. `msg` is just a value. It can't throw an exception. Did you mean to use on `Marshal.from_channel`?

Reply:
Ouch, nice catch @glennsl ! My intention was to have a `try...catch` around the Marshal call, since that can potentially fail (I believe) and at least log it.

Fixed by wrapping in a `try` now